lkl: Add LKL_HIJACK_SINGLE_CPU

liuyuan10 opened the pull request:
This option pins all LKL kernel threads onto single cpus. It's better
than taskset in the sense that it doesn't require how the application is
run.

Signed-off-by: Yuan Liu liuyuan@google.com

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lkl/linux/177)

<!-- Reviewable:end -->

thehajime commented:
thanks for another great patch - I will review and try it and get you back later.

> taskset offers better TCP_RR of course, but a bit lower TCP_STREAM vs this commit. It's because the network polling threads are not pinned.

I'm not well understanding  your above comment.  Didn't `taskset` stick processor(s) for the tx/rx poliing threads of virtio-net ?

liuyuan10 commented:
My words may be confusing. taskset pins polling threads while this PR doesn't. That's why this PR can provides tiny improvement. The polling threads can run in parallel with LKL threads

thehajime commented:
thanks @liuyuan10 for the answer.

why doesn't this PR pin the polling threads ? after commented out the following block, it should pin the polling threads too ? 

```
        /* restore cpu affinity */
//      if (should_pin_cpu)
//              PinToCpus(&ori_cpu);

```

liuyuan10 commented:
It depends on where PinToFirstCpu is called. Children threads inherit cpu affinity from parent. Polling thread is created in lkl_net_dev_add.

I added two single_cpu_mode to decide if pining polling threads. The performance results I showed before is wrong due to bad runtime lib. I updated it in commit msg:
branch TCP_RR(us) TCP_STREAM(Mb/s)
master 95 409
taskset 65 696
mode1 92 725
mode2 82 645

mode1 can better TCP_STREAM but worse TCP_RR.

thehajime commented:
thanks for the clarification.

What I was expecting is that this PR does the same thing as taskset does.
Is that your intention ?

If I changed like below, the result of TCP_RR looks similar, but TCP_STREAM is different (taskset is still better than this PR).

```
diff --git a/tools/lkl/lib/hijack/init.c b/tools/lkl/lib/hijack/init.c
index 41f56aa9710f..7b1618b9c32f 100644
--- a/tools/lkl/lib/hijack/init.c
+++ b/tools/lkl/lib/hijack/init.c
@@ -229,6 +229,11 @@ hijack_init(void)
                        should_pin_cpu = 0;
                }
        }
+       /* Pin to a single cpu before calling lkl_start_kernel so that all lkl
+        * kernel threads will be pinned to the same cpu.
+        */
+       if (should_pin_cpu)
+               PinToFirstCpu(&ori_cpu);
 
        /* Must be run before lkl_netdev_tap_create */
        fixup_netdev_linux_fdnet_ops();
@@ -277,11 +282,6 @@ hijack_init(void)
        else
                lkl_register_dbg_handler();
 
-       /* Pin to a single cpu before calling lkl_start_kernel so that all lkl
-        * kernel threads will be pinned to the same cpu.
-        */
-       if (should_pin_cpu)
-               PinToFirstCpu(&ori_cpu);
 
        ret = lkl_start_kernel(&lkl_host_ops, 64 * 1024 * 1024, "");
        if (ret) {
@@ -290,8 +290,8 @@ hijack_init(void)
        }
 
        /* restore cpu affinity */
-       if (should_pin_cpu)
-               PinToCpus(&ori_cpu);
+//     if (should_pin_cpu)
+//             PinToCpus(&ori_cpu);
 
        /* fillup FDs up to LKL_FD_OFFSET */
        ret = lkl_sys_mknod("/dev_null", LKL_S_IFCHR | 0600, LKL_MKDEV(1, 3));
```

liuyuan10 commented:
I don't know which thread runs the LD_PRELOAD. If it's the same thread running main(), then your change should have exact same behaviour as task set. 

I can observe similar TCP_STREAM with taskset with your change. Make sure you use "taskset -c 0" so that they are pinned to the same core.

The intention of this PR is not to substitute taskset. It's to provide improvement when you can't use taskset. For example, you have a large application and you don't want to just pin every thread on a single core.

thehajime commented:
> I don't know which thread runs the LD_PRELOAD. If it's the same thread running main(), then your change should have exact same behaviour as task set.

opurdila commented on tools/lkl/lib/hijack/init.c:
```diff
@@ -177,6 +180,26 @@ void fixup_netdev_linux_fdnet_ops(void)
+	}
+}
+
+static void PinToFirstCpu(const cpu_set_t* cpus) {
```
Minor thing regarding coding style, it would be great if you could you fix up this and the function above.

liuyuan10 replied:
Thanks for reminding. Fixed.
